image/image.go

```diff
@@ -0,0 +1,386 @@
+	if osutil.FileExists(snapName) {
+		return copyLocalSnapFile(snapName, dlOpts.TargetDir)
+	}
+
+	return downloadSnapWithSideInfo(snapName, dlOpts)
+}
+
+func bootstrapToRootdir(opts *Options) error {
+	if opts.Rootdir != "" {
+		dirs.SetRootDir(opts.Rootdir)
+		defer dirs.SetRootDir("/")
+	}
+
+	// sanity check target
+	if osutil.FileExists(dirs.SnapStateFile) {
+		return fmt.Errorf("cannot bootstrap over existing system")
+	}
+
+	model, err := decodeModelAssertion(opts.ModelAssertionFn)
+	if err != nil {
+		return err
+	}
+
+	// put snaps in place
+	if err := os.MkdirAll(dirs.SnapBlobDir, 0755); err != nil {
+		return err
+	}
+
+	snapSeedDir := filepath.Join(dirs.SnapSeedDir, "snaps")
+	dlOpts := &downloadOptions{
+		TargetDir:    snapSeedDir,
+		Channel:      opts.Channel,
+		StoreID:      model.Store(),
+		Architecture: model.Architecture(),
+	}
+
+	// FIXME: support sideloading snaps by copying the boostrap.snaps
+	//        first and keeping track of the already downloaded names
+	snaps := []string{}
+	snaps = append(snaps, opts.Snaps...)
+	snaps = append(snaps, model.Gadget())
+	snaps = append(snaps, model.Core())
+	snaps = append(snaps, model.Kernel())
+	snaps = append(snaps, model.RequiredSnaps()...)
+
+	for _, d := range []string{dirs.SnapBlobDir, snapSeedDir} {
+		if err := os.MkdirAll(d, 0755); err != nil {
+			return err
+		}
+	}
+
+	var seedYaml snap.Seed
+	for _, snapName := range snaps {
+		fmt.Printf("Fetching %s\n", snapName)
+		fn, info, err := acquireSnap(snapName, dlOpts)
+		if err != nil {
+			return err
+		}
+		// kernel/os are required for booting
+		if snapName == model.Kernel() || snapName == model.Core() {
+			if err := osutil.CopyFile(fn, filepath.Join(dirs.SnapBlobDir, filepath.Base(fn)), 0); err != nil {
+				return err
+			}
+		}
+		// set seed.yaml
+		seedYaml.Snaps = append(seedYaml.Snaps, &snap.SeedSnap{
+			Name:        info.Name(),
+			SnapID:      info.SnapID,
+			Revision:    info.Revision,
+			Channel:     info.Channel,
+			DeveloperID: info.DeveloperID,
+			Developer:   info.Developer,
+			File:        filepath.Base(fn),
+		})
+	}
+	if err := seedYaml.Write(); err != nil {
+		return fmt.Errorf("cannot write seed.yaml: %s", err)
+	}
+
+	// now do the bootloader stuff
+	if err := partition.InstallBootConfig(opts.GadgetUnpackDir); err != nil {
+		return err
+	}
+
+	if err := setBootvars(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setBootvars() error {
+	// set the bootvars for kernel/os snaps so that the system
+	// actually boots and can do the `firstboot` import of the snaps.
+	//
+	// there is also no mounted os/kernel snap, all we have are the
+	// blobs
+
+	bootloader, err := partition.FindBootloader()
+	if err != nil {
+		return fmt.Errorf("can not set kernel/os bootvars: %s", err)
+	}
+
+	snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDir, "*.snap"))
+	if len(snaps) == 0 {
+		return fmt.Errorf("internal error: cannot find os/kernel snap")
+	}
+	for _, fullname := range snaps {
```

Comment:
Yeah, that was my intention. But I agree that fullname is a bit confusing because one wonders what it tries to convey. I changed it to `fn` now which is either filename or fullname and should carry less cognitive burden.